File: smooth_data.m
function smoothedSignal = smooth_data(signalIn, windowSize)

coder.inline('never')

% signal length
n = length(signalIn);
% use half window for boudary situation
halfWin = floor(windowSize / 2);
% initialization
smoothedSignal = zeros(size(signalIn));

% Initialize sum for the first window considering boundary conditions
windowSum = sum(signalIn(1:halfWin));
count = halfWin; % Actual count of elements in the initial window

% Initialize first smoothed value especially if it's within a boundary condition
for i = 1:min(halfWin, n)
    smoothedSignal(i) = windowSum / count;
    % Only add new elements if we are not at the end
    if i + halfWin <= n
        windowSum = windowSum + signalIn(i + halfWin);
        count = count + 1;
    end
end

% Smooth for every point after the initial part
for i = (halfWin + 1):(n - halfWin)
    % Add the next item in the window and remove the first one
    windowSum = windowSum + signalIn(i + halfWin) - signalIn(i - halfWin);
    % The count remains constant in the middle part
    smoothedSignal(i) = windowSum / windowSize; % Here the window is always full-sized
end

% Handle the end boundary where the window shrinks again
for i = max(n - halfWin + 1, halfWin + 1):n
    % Only subtract elements if we are not at the beginning
    if i - halfWin > 1
        windowSum = windowSum - signalIn(i - halfWin - 1);
        count = count - 1;
    end
    smoothedSignal(i) = windowSum / count;
end
end

File: codegen/ppg_process0422/smooth_data.c
/*
 * Academic License - for use in teaching, academic research, and meeting
 * course requirements at degree granting institutions only.  Not for
 * government, commercial, or other organizational use.
 *
 * smooth_data.c
 *
 * Code generation for function 'smooth_data'
 *
 */

/* Include files */
#include "smooth_data.h"
#include "rt_nonfinite.h"

/* Function Definitions */
void smooth_data(const float signalIn[150], double smoothedSignal[150])
{
  float windowSum;
  int count;
  int k;
  /*  signal length */
  /*  use half window for boudary situation */
  /*  initialization */
  /*  Initialize sum for the first window considering boundary conditions */
  windowSum = signalIn[0];
  for (k = 0; k < 24; k++) {
    windowSum += signalIn[k + 1];
  }
  count = 25;
  /*  Actual count of elements in the initial window */
  /*  Initialize first smoothed value especially if it's within a boundary
   * condition */
  for (k = 0; k < 25; k++) {
    smoothedSignal[k] = windowSum / (float)count;
    /*  Only add new elements if we are not at the end */
    windowSum += signalIn[k + 25];
    count++;
  }
  /*  Smooth for every point after the initial part */
  for (k = 0; k < 100; k++) {
    /*  Add the next item in the window and remove the first one */
    windowSum = (windowSum + signalIn[k + 50]) - signalIn[k];
    /*  The count remains constant in the middle part */
    smoothedSignal[k + 25] = windowSum / 50.0F;
    /*  Here the window is always full-sized */
  }
  /*  Handle the end boundary where the window shrinks again */
  for (k = 0; k < 25; k++) {
    /*  Only subtract elements if we are not at the beginning */
    windowSum -= signalIn[k + 99];
    count--;
    smoothedSignal[k + 125] = windowSum / (float)count;
  }
}

/* End of code generation (smooth_data.c) */
